### PR TITLE
handle imbalanced datasets

### PR DESCRIPTION
**Pull request by prithvikannan**
Signed-off-by: Prithvi Kannan <prithvi.kannan@databricks.com>

## Related Issues/PRs

 #xxx

## What changes are proposed in this pull request?

Handle imbalanced datasets in the classification pipeline. If n_rows in the training dataset is >5000, downsample the majority class to produce a 70:30 ratio. 

Users can override this ratio in the `pipeline.yaml`. Example for how to produce a 50:50 ratio.
```
    train:
        resampling_minority_percentage: 0.5
```

Rebalancing is by default enabled. Users can also disable rebalancing in the `pipeline.yaml`. 
```
    train:
        rebalancing_enabled: False
```

Note that we must pass `class_weight` to the estimator function to restore the original probability distribution of the train dataset, otherwise we would be making the assumption that the rebalanced distribution is the true distribution.

## How is this patch tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests (describe details, including test results, below)

* Testing on small datasets (75 examples)
image

* Testing on (relatively) balanced datasets
image

* Testing on imbalanced datasets. NOTE: I disabled the min threshold for this test case
image

* Testing on imbalanced datasets with `resampling_minority_percentage=0.5` . NOTE: I disabled the min threshold for this test case
image

## Does this PR change the documentation?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly in the documentation preview.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/recipes/steps/train.py [hunk 1]
```diff
@@ -149,6 +152,7 @@ def _validate_and_apply_step_config(self):
                 error_code=INVALID_PARAMETER_VALUE,
             )
         self.positive_class = self.step_config.get("positive_class")
+        self.rebalancing_enabled = self.step_config.get("rebalancing_enabled", True)
```
Comment (jinzhang21) [inline mlflow/recipes/steps/train.py:155]: Prefer to use a verb, e.g. `use_balanced_training_data` or `rebalance_training_data`

File: mlflow/recipes/steps/train.py [hunk 2]
```diff
@@ -1121,8 +1138,64 @@ def _safe_dump_with_numeric_values(self, data, file, **kwargs):
...
+        df_negative_class = train_df[train_df[self.target_col] != self.positive_class]
+
+        classes = np.unique(train_df[self.target_col])
+        class_weights = compute_class_weight(
+            class_weight="balanced",
+            classes=classes,
+            y=train_df[self.target_col],
+        )
+
+        self.original_class_weights = dict(zip(classes, class_weights))
```
Comment (jinzhang21) [inline mlflow/recipes/steps/train.py:1168]: Can we factor this section out, compute class weights irrespective whether rebalance is enabled? It's a useful metric to show in step card regardless.

File: mlflow/recipes/steps/train.py [hunk 3]
```diff
@@ -465,6 +479,11 @@ def _get_user_defined_estimator(self, X_train, y_train, validation_df, run, outp
         sys.path.append(self.recipe_root)
         estimator_fn = getattr(importlib.import_module(train_module_name), estimator_method_name)
         estimator_hardcoded_params = self.step_config["estimator_params"]
+
+        # if using rebalancing pass in original class weights to preserve original distribution
+        if self.using_rebalancing:
+            estimator_hardcoded_params["class_weight"] = self.original_class_weights
```
Comment (jinzhang21) [inline mlflow/recipes/steps/train.py:485]: Reading the `class_weight` definition, this line seems to be no-op since the `class_weights` are computed in exactly the same way in `compute_class_weight` with `balanced` mode.
https://scikit-learn.org/stable/modules/generated/sklearn.utils.class_weight.compute_class_weight.html
https://scikit-learn.org/stable/modules/generated/sklearn.ensemble.RandomForestClassifier.html

File: mlflow/recipes/steps/train.py [hunk 4]
```diff
@@ -1121,8 +1138,64 @@ def _safe_dump_with_numeric_values(self, data, file, **kwargs):
...
+            class_weight="balanced",
+            classes=classes,
+            y=train_df[self.target_col],
+        )
+
+        self.original_class_weights = dict(zip(classes, class_weights))
+
+        if len(df_positive_class) > len(df_negative_class):
+            df_minority_class, df_majority_class = df_negative_class, df_positive_class
+            original_minority_percentage = len(df_negative_class) / len(train_df)
```
Comment (jinzhang21) [inline mlflow/recipes/steps/train.py:1172]: Can we compute this always as `original_minority_percentage = len(df_minority_class) / len(train_df)`?

File: mlflow/recipes/steps/train.py [hunk 5]
```diff
@@ -1121,8 +1138,64 @@ def _safe_dump_with_numeric_values(self, data, file, **kwargs):
...
+        if original_minority_percentage >= resampling_minority_percentage:
+            _logger.info(
+                f"Class imbalance of {original_minority_percentage:.2f} "
+                f"is better than {resampling_minority_percentage}, no need to rebalance"
+            )
+            return train_df
+
+        _logger.info(
+            f"Detected class imbalance: original count - minority class: "
+            f"{len(df_minority_class)} vs majority class: {len(df_majority_class)}"
```
Comment (jinzhang21) [inline mlflow/recipes/steps/train.py:1186]: Showing the ratio as well, which is probably more useful in this context.

File: mlflow/recipes/steps/train.py [hunk 6]
```diff
@@ -1121,8 +1138,64 @@ def _safe_dump_with_numeric_values(self, data, file, **kwargs):
...
+        majority_class_target = int(
+            len(df_minority_class)
+            * (1 - resampling_minority_percentage)
+            / resampling_minority_percentage
+        )
+        df_majority_downsampled = df_majority_class.sample(majority_class_target)
+        _logger.info(
+            f"After downsampling: original count - minority class: "
+            f"{len(df_minority_class)} vs majority class: {len(df_majority_downsampled)}"
+        )
```
Comment (jinzhang21) [inline mlflow/recipes/steps/train.py:1198]: Same here. Show the ratio after rebalancing.

File: mlflow/recipes/steps/train.py [hunk 7]
```diff
@@ -1121,8 +1138,64 @@ def _safe_dump_with_numeric_values(self, data, file, **kwargs):
...
+        train_df = pd.concat([df_minority_class, df_majority_downsampled], axis=0)
```
Comment (jinzhang21) [inline mlflow/recipes/steps/train.py:1199]: This is dangerous. Should `shuffle` to make sure the SGD optimizer does't go crazy over one class for long period of time:)

**Comment by prithvikannan** [comment]
Updated logging:
image

File: mlflow/recipes/steps/train.py (hunk 3 above)
Comment (bbarnes52) [reply mlflow/recipes/steps/train.py:495]: I could be wrong, but I believe this implementation is slightly different. The original class distribution is not known to the estimator since rebalancing is done upstream.

**Review by bbarnes52 — APPROVED** [review]
Excellent work. 

Fwiw, I'm not entirely sold on rebalancing by default. That seems appropriate for AutoML but MLP has so far given the user full control over what transformations are performed on their data. If it were me, I'd let the user decide, and display a warning if their data is highly imbalanced to suggest using rebalancing.

That said, I don't feel strongly and am supportive of landing this as-is.

File: mlflow/recipes/steps/train.py (hunk 3 above)
Comment (jinzhang21) [reply mlflow/recipes/steps/train.py:495]: Yes. @bbarnes52 You are right. The weights have changed effectively at this point.

**Comment by jinzhang21** [comment]
I didn't get a chance to comment that there's little unit test added to ensure the new functionality will actually work as intended and remain so against future changes. Consider file a ticket to track it as a fixit. @prithvikannan
